File: common/src/main/java/juuxel/adorn/block/AbstractTableBlock.java
package juuxel.adorn.block;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.util.RandomSource;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.ScheduledTickAccess;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.SimpleWaterloggedBlock;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.level.block.state.properties.BooleanProperty;
import net.minecraft.world.level.material.FluidState;
import net.minecraft.world.level.material.Fluids;
import net.minecraft.world.level.pathfinder.PathComputationType;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.VoxelShape;

public abstract class AbstractTableBlock extends CarpetedBlock implements SimpleWaterloggedBlock {
    public static final BooleanProperty NORTH = BlockStateProperties.NORTH;
    public static final BooleanProperty EAST = BlockStateProperties.EAST;
    public static final BooleanProperty SOUTH = BlockStateProperties.SOUTH;
    public static final BooleanProperty WEST = BlockStateProperties.WEST;
    public static final BooleanProperty WATERLOGGED = BlockStateProperties.WATERLOGGED;

    public AbstractTableBlock(Properties settings) {
        super(settings);
        registerDefaultState(defaultBlockState().setValue(WATERLOGGED, false));
    }

    protected abstract boolean canConnectTo(BlockState state, Direction sideOfSelf);

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        super.createBlockStateDefinition(builder);
        builder.add(NORTH, EAST, SOUTH, WEST, WATERLOGGED);
    }

    @Override
    public BlockState getStateForPlacement(BlockPlaceContext ctx) {
        return updateConnections(
            super.getStateForPlacement(ctx)
                .setValue(WATERLOGGED, ctx.getLevel().getFluidState(ctx.getClickedPos()).getType() == Fluids.WATER),
            ctx.getLevel(),
            ctx.getClickedPos()
        );
    }

    @Override
    public FluidState getFluidState(BlockState state) {
        return state.getValue(WATERLOGGED) ? Fluids.WATER.getSource(false) : super.getFluidState(state);
    }

    @Override
    protected BlockState updateShape(BlockState state, LevelReader world, ScheduledTickAccess tickView, BlockPos pos, Direction direction, BlockPos neighborPos, BlockState neighborState, RandomSource random) {
        if (state.getValue(WATERLOGGED)) {
            tickView.scheduleTick(pos, Fluids.WATER, Fluids.WATER.getTickDelay(world));
        }

        return updateConnections(super.updateShape(state, world, tickView, pos, direction, neighborPos, neighborState, random), world, pos);
    }

    private BlockState updateConnections(BlockState state, BlockGetter world, BlockPos pos) {
        return state.setValue(NORTH, canConnectTo(world.getBlockState(pos.relative(Direction.NORTH)), Direction.NORTH))
            .setValue(EAST, canConnectTo(world.getBlockState(pos.relative(Direction.EAST)), Direction.EAST))
            .setValue(SOUTH, canConnectTo(world.getBlockState(pos.relative(Direction.SOUTH)), Direction.SOUTH))
            .setValue(WEST, canConnectTo(world.getBlockState(pos.relative(Direction.WEST)), Direction.WEST));
    }

    @Override
    public VoxelShape getShape(BlockState state, BlockGetter world, BlockPos pos, CollisionContext context) {
        return getShapeForKey(
            getShapeKey(
                state.getValue(NORTH), state.getValue(EAST), state.getValue(SOUTH), state.getValue(WEST),
                isCarpetingEnabled() && state.getValue(CARPET).isPresent()
            )
        );
    }

    protected static int getShapeKey(boolean north, boolean east, boolean south, boolean west, boolean hasCarpet) {
        return (north ? 1 : 0) << 4 | (east ? 1 : 0) << 3 | (south ? 1 : 0) << 2 | (west ? 1 : 0) << 1 | (hasCarpet ? 1 : 0);
    }

    protected abstract VoxelShape getShapeForKey(int key);

    @Override
    public boolean isPathfindable(BlockState state, PathComputationType type) {
        return false;
    }
}

File: src/main/kotlin/juuxel/adorn/block/AbstractTableBlock.kt
@file:Suppress("DEPRECATION")
package juuxel.adorn.block

import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.ShapeContext
import net.minecraft.block.Waterloggable
import net.minecraft.entity.ai.pathing.NavigationType
import net.minecraft.fluid.Fluids
import net.minecraft.item.ItemPlacementContext
import net.minecraft.state.StateManager
import net.minecraft.state.property.Properties
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.util.shape.VoxelShape
import net.minecraft.world.BlockView
import net.minecraft.world.WorldAccess

abstract class AbstractTableBlock(settings: Settings) : CarpetedBlock(settings), Waterloggable {
    init {
        defaultState = defaultState.with(WATERLOGGED, false)
    }

    protected abstract fun canConnectTo(state: BlockState, sideOfSelf: Direction): Boolean

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        super.appendProperties(builder)
        builder.add(NORTH, EAST, SOUTH, WEST, WATERLOGGED)
    }

    override fun getPlacementState(context: ItemPlacementContext): BlockState =
        updateConnections(
            super.getPlacementState(context)!!
                .with(Properties.WATERLOGGED, context.world.getFluidState(context.blockPos).fluid == Fluids.WATER),
            context.world,
            context.blockPos
        )

    override fun getFluidState(state: BlockState) =
        if (state[Properties.WATERLOGGED]) Fluids.WATER.getStill(false)
        else super.getFluidState(state)

    override fun getStateForNeighborUpdate(
        state: BlockState,
        direction: Direction,
        neighborState: BlockState,
        world: WorldAccess,
        pos: BlockPos,
        neighborPos: BlockPos
    ) = updateConnections(
        super.getStateForNeighborUpdate(state, direction, neighborState, world, pos, neighborPos),
        world,
        pos
    )

    private fun updateConnections(
        state: BlockState,
        world: WorldAccess,
        pos: BlockPos
    ) = state.with(NORTH, canConnectTo(world.getBlockState(pos.offset(Direction.NORTH)), Direction.NORTH))
        .with(EAST, canConnectTo(world.getBlockState(pos.offset(Direction.EAST)), Direction.EAST))
        .with(SOUTH, canConnectTo(world.getBlockState(pos.offset(Direction.SOUTH)), Direction.SOUTH))
        .with(WEST, canConnectTo(world.getBlockState(pos.offset(Direction.WEST)), Direction.WEST))

    override fun getOutlineShape(state: BlockState, view: BlockView?, pos: BlockPos?, context: ShapeContext?) =
        getShapeForKey(
            Bits.buildTableState(
                state[NORTH], state[EAST], state[SOUTH], state[WEST],
                isCarpetingEnabled() && state[CARPET].isPresent
            )
        )

    protected abstract fun getShapeForKey(key: Byte): VoxelShape

    override fun canPathfindThrough(state: BlockState, world: BlockView, pos: BlockPos, type: NavigationType) = false

    companion object {
        val NORTH = Properties.NORTH
        val EAST = Properties.EAST
        val SOUTH = Properties.SOUTH
        val WEST = Properties.WEST
        val CARPET = CarpetedBlock.CARPET
        val WATERLOGGED = Properties.WATERLOGGED
    }
}
